--- dwm.c	2012-07-31 12:46:36.009800225 -0400
+++ /home/via/dev/dwm-5.8.2/dwm.c	2012-07-31 12:48:55.883126301 -0400
@@ -684,6 +684,8 @@ drawbar(Monitor *m) {
 	unsigned long *col;
 	Client *c;
 
+	unsigned int num_cl = 0;
+
 	for(c = m->clients; c; c = c->next) {
 		occ |= c->tags;
 		if(c->isurgent)
@@ -713,15 +715,24 @@ drawbar(Monitor *m) {
 	}
 	else
 		dc.x = m->ww;
+
+	for (c = m->clients; c != NULL; c = c->next) {
+		if (ISVISIBLE(c)) num_cl++;
+	}
+
 	if((dc.w = dc.x - x) > bh) {
 		dc.x = x;
-		if(m->sel) {
-			col = m == selmon ? dc.sel : dc.norm;
-			drawtext(m->sel->name, col, False);
-			drawsquare(m->sel->isfixed, m->sel->isfloating, False, col);
-		}
-		else
+		if (num_cl) {
+			dc.w = dc.w / num_cl;
+			for (c = m->clients; c != NULL; c = c->next) {
+				col = ((m == selmon) && (m->sel == c)) ? dc.sel : dc.norm;
+				drawtext(c->name, col, False);
+				drawsquare(c->isfixed, c->isfloating, False, col);
+				dc.x += dc.w;
+			}
+		} else {
 			drawtext(NULL, dc.norm, False);
+		}
 	}
 	XCopyArea(dpy, dc.drawable, m->barwin, dc.gc, 0, 0, m->ww, bh, 0, 0);
 	XSync(dpy, False);
